Check cluster existence/Don't reconcile if deleted

hryan-ib opened the pull request:
This adds a check to see if the cluster was deleted outside of the operator realm. If the cluster was deleted and it is enqueued for an update, it will revert back to Pending stage and remake the cluster. IF the cluster is not found but the CR is trying to be deleted, it will note that the cluster is already gone and delete the CR.

This also prevents the CR from being reconciled again on its delete event.

hryan-ib committed:
delete working

hryan-ib committed:
clean up

hryan-ib committed:
dont reconcile on delete

drewwells commented:
fixes #23 #17

bjeevan-ib commented on pkg/controller/cluster/cluster_controller.go:
```diff
@@ -201,15 +224,20 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
+		err := k.GetCluster(instance.Spec.KopsConfig)
 		if err != nil {
-			// FIXME - Ensure that delete implementation is idempotent and safe to invoke multiple times.
-			// If we call delete and the cluster is not present it will cause error and it will keep erroring out
-			//return reconcile.Result{}, err
+			reqLogger.Info("Cluster is already deleted...")
```
Can we not explicitly check if cluster is present or not (instead of err != nil)?

seizadi replied:
@hryan-ib 
I think that is fair, we could change k.GetCluster() to return (bool, error)
Then the code would be: 
found, err := k.GetCluser()

drewwells replied:
Include the error in the log message

hryan-ib replied:
I have updated the code to use a boolean. It will check that the error is for not finding the cluster ("exit status 1") and will report that as a cluster not found. Any other error will be treated as a command error and will exit reconcile. If cluster found, the boolean is true and it will continue reconciling

drewwells commented on pkg/controller/cluster/cluster_controller.go:
```diff
@@ -125,14 +135,27 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
+		if instance.Status.Phase != clusteroperatorv1alpha1.ClusterPending && instance.Status.Phase != "" {
+			err := k.GetCluster(instance.Spec.KopsConfig)
+			if err != nil {
+				reqLogger.Info("Cluster does not exist... resetting to PENDING")
```
We should explicitly be looking for a specific type of error here too. `strings.Contains(err.Error(), "some useful message")` or log the error to do it later `reqLogger.WithValues("error", err).Info("....")`.

hryan-ib replied:
Looks like Kops only sends out the error as "exit status 1" so its not very descriptive. I have updated the GetCluster function to catch anything with "exit status 1" and report that as a cluster not existing. Any other error will be an actual command error and will exit from reconcile. The command streaming will also print out if the cluster is not found

drewwells commented on pkg/controller/cluster/cluster_controller.go:
```diff
@@ -201,15 +224,20 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
-			// FIXME - Ensure that delete implementation is idempotent and safe to invoke multiple times.
-			// If we call delete and the cluster is not present it will cause error and it will keep erroring out
-			//return reconcile.Result{}, err
+			reqLogger.Info("Cluster is already deleted...")
```
```suggestion
			reqLogger.WithValues("error", err).Info("Cluster is already deleted...")
```
Or relevant syntax to this logger

drewwells reviewed (changes requested):
include errors in logs, it will help debug edge cases

hryan-ib committed:
fixing getcluster error logging

hryan-ib commented:
I addressed the comments on this PR. @seizadi @drewwells any qualms with merging this in? Although I am still not sure what causes the checks to fail...

hryan-ib committed:
Merge branch 'master' of https://github.com/infobloxopen/cluster-operator into delete

hryan-ib committed:
removing added bindata file in kops vendor
